File: apps/mo1-ios/src/push.ts
// push — APNs payload -> deeplink route + badge (design §2-2 PushKit, test §7
// Push). The frozen mobile.MobilePushPayload keeps { title, body, data? }; the
// routing/badge/type fields the design lists travel inside `data` (string map,
// APNs custom keys). We read them defensively and never crash on a missing key.
import type { mobile } from "@dub/types";
import { parseDeepLink, type Route } from "./deeplink.js";

export interface ParsedPush {
  title: string;
  body: string;
  route: Route; // where a tap navigates (home if no deepLink)
  badge: number | null; // app-icon badge (null = leave unchanged)
  notificationId: string | null;
  type: string | null; // notification.NotificationType (open vocabulary)
  correlationId: string | null;
}

function num(v: string | undefined): number | null {
  if (v === undefined) return null;
  const n = Number(v);
  return Number.isFinite(n) ? n : null;
}

/** Interpret an incoming APNs payload; total (never throws). */
export function parsePush(payload: mobile.MobilePushPayload): ParsedPush {
  const data = payload.data ?? {};
  const deepLink = data.deepLink;
  return {
    title: payload.title,
    body: payload.body,
    route: deepLink ? parseDeepLink(deepLink) : { name: "home", params: {} },
    badge: num(data.badge),
    notificationId: data.notificationId ?? null,
    type: data.type ?? null,
    correlationId: data.correlationId ?? null,
  };
}

File: apps/mo1-ios/ios/Sources/Mo1Core/Domain/Push.swift
// Push — APNs payload -> deeplink route + badge (design §2-2 PushKit, test §7
// Push), mirrors push.ts. Routing/badge/type fields travel inside `data` (APNs
// custom string keys); read defensively and never crash on a missing key.
import Foundation

public struct ParsedPush: Equatable, Sendable {
    public var title: String
    public var body: String
    public var route: Route              // where a tap navigates (home if no deepLink)
    public var badge: Int?               // app-icon badge (nil = leave unchanged)
    public var notificationId: String?
    public var type: String?             // NotificationType (open vocabulary)
    public var correlationId: String?
}

private func numericBadge(_ value: String?) -> Int? {
    guard let value, let d = Double(value), d.isFinite else { return nil }
    return Int(d)
}

/// Interpret an incoming APNs payload; total (never throws).
public func parsePush(_ payload: MobilePushPayload) -> ParsedPush {
    let data = payload.data ?? [:]
    let deepLink = data["deepLink"]
    return ParsedPush(
        title: payload.title,
        body: payload.body,
        route: deepLink.map(parseDeepLink) ?? Route(name: .home),
        badge: numericBadge(data["badge"]),
        notificationId: data["notificationId"],
        type: data["type"],
        correlationId: data["correlationId"]
    )
}
